File: typecast-js/src/types/Voices.ts
import { TTSModel } from './TextToSpeech';

/**
 * V1 Voices response (deprecated, use VoiceV2Response instead)
 */
export interface VoicesResponse {
  voice_id: string;
  voice_name: string;
  model: TTSModel;
  emotions: string[];
}

/**
 * Gender classification for voices
 */
export type GenderEnum = 'male' | 'female';

/**
 * Age group classification for voices
 */
export type AgeEnum = 'child' | 'teenager' | 'young_adult' | 'middle_age' | 'elder';

/**
 * Use case categories for voices
 */
export type UseCaseEnum =
  | 'Announcer'
  | 'Anime'
  | 'Audiobook'
  | 'Conversational'
  | 'Documentary'
  | 'E-learning'
  | 'Rapper'
  | 'Game'
  | 'Tiktok/Reels'
  | 'News'
  | 'Podcast'
  | 'Voicemail'
  | 'Ads';

/**
 * Model information with supported emotions
 */
export interface ModelInfo {
  /** TTS model version (e.g., ssfm-v21, ssfm-v30) */
  version: TTSModel;
  /** List of supported emotions for this model */
  emotions: string[];
}

/**
 * V2 Voice response with enhanced metadata
 */
export interface VoiceV2Response {
  /** Unique voice identifier */
  voice_id: string;
  /** Human-readable name of the voice */
  voice_name: string;
  /** List of supported TTS models with their available emotions */
  models: ModelInfo[];
  /** Voice gender classification */
  gender?: GenderEnum | null;
  /** Voice age group classification */
  age?: AgeEnum | null;
  /** List of use case categories this voice is suitable for */
  use_cases?: string[];
}

/**
 * Filter options for V2 voices endpoint
 */
export interface VoicesV2Filter {
  /** Filter by TTS model */
  model?: TTSModel;
  /** Filter by gender */
  gender?: GenderEnum;
  /** Filter by age group */
  age?: AgeEnum;
  /** Filter by use case */
  use_cases?: UseCaseEnum;
}

File: typecast-python/src/typecast/models/voices.py
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field

from .tts import TTSModel


class VoicesResponse(BaseModel):
    """V1 Voices response (deprecated, use VoiceV2Response instead)"""

    voice_id: str
    voice_name: str
    model: str
    emotions: list[str]


class GenderEnum(str, Enum):
    """Gender classification for voices"""

    MALE = "male"
    FEMALE = "female"


class AgeEnum(str, Enum):
    """Age group classification for voices"""

    CHILD = "child"
    TEENAGER = "teenager"
    YOUNG_ADULT = "young_adult"
    MIDDLE_AGE = "middle_age"
    ELDER = "elder"


class UseCaseEnum(str, Enum):
    """Use case categories for voices"""

    ANNOUNCER = "Announcer"
    ANIME = "Anime"
    AUDIOBOOK = "Audiobook"
    CONVERSATIONAL = "Conversational"
    DOCUMENTARY = "Documentary"
    E_LEARNING = "E-learning"
    RAPPER = "Rapper"
    GAME = "Game"
    TIKTOK_REELS = "Tiktok/Reels"
    NEWS = "News"
    PODCAST = "Podcast"
    VOICEMAIL = "Voicemail"
    ADS = "Ads"


class ModelInfo(BaseModel):
    """Model information with supported emotions"""

    version: TTSModel
    emotions: list[str]


class VoiceV2Response(BaseModel):
    """V2 Voice response with enhanced metadata"""

    voice_id: str
    voice_name: str
    models: list[ModelInfo]
    gender: Optional[GenderEnum] = None
    age: Optional[AgeEnum] = None
    use_cases: Optional[list[str]] = None


class VoicesV2Filter(BaseModel):
    """Filter options for V2 voices endpoint"""

    model: Optional[TTSModel] = None
    gender: Optional[GenderEnum] = None
    age: Optional[AgeEnum] = None
    use_cases: Optional[UseCaseEnum] = None


class CustomVoice(BaseModel):
    """Quick-cloned custom voice returned by `POST /v1/voices/clone`.

    Attributes:
        voice_id: Custom voice identifier with `uc_` prefix.
            Use this value as `voice_id` in `text_to_speech` / `text_to_speech_with_timestamps`.
        name: Human-readable name (1-30 chars).
        model: Engine model the voice was cloned for (`ssfm-v21` or `ssfm-v30`).
    """

    voice_id: str = Field(..., description="Custom voice identifier (uc_ prefix)")
    name: str = Field(..., description="Human-readable voice name")
    model: str = Field(..., description="Engine model: ssfm-v21 or ssfm-v30")
